### PR TITLE
ci: support staging for all branches but main

### PR DESCRIPTION
**Pull request by v1v**
### What

Produce snapshot and releasable artifacts, for such:

* it publishes packages after every commit to all active release branch
* it always builds snapshot images off the main branch
* it always builds staging images off other branches

### Implementation details

- Matrix stage to package for `staging` and `snapshot` 
- DRA stage to :
  - First one to combine all the packages/binaries
  - Second one to run the snapshot release workflow
  - Third one, only other branches but main, to run the staging release workflow

### Issue

Follow up from https://github.com/elastic/apm-server/pull/7530

### Question

The existing implementation uses `release-manager-snapshot` -> https://github.com/elastic/apm-server/blob/27d244ea6b88edb69f38165ab86a7a69892f0676/.ci/packaging.groovy#L93 I guess `staging` should not use this but `release-manager-release`, if so:

1) ~~We need to run twice the packaging for the release branches, one with the `snapshot` and another one with `images`, am I right?~~

### CODE REVIEW

**Comment by axw** [comment]
> We need to run twice the packaging for the release branches, one with the snapshot and another one with images, am I right?

I think we just start building staging (BC) images as soon as we've created a release branch at feature freeze time. Can you please check with the unified release team?

**Review by axw — COMMENTED** [review]
Looks good! IIANM we should also remove SNAPSHOT from environment at the top of packaging.groovy

**Comment by v1v** [comment]
> I think we just start building staging (BC) images as soon as we've created a release branch at feature freeze time. Can you please check with the unified release team?

I'll double check with the release team, though I need to figure out how this particular time event can be discovered within the CI

File: .ci/packaging.groovy [hunk 1]
```diff
@@ -127,17 +109,9 @@ pipeline {
...
-              dockerLogin(secret: env.DOCKER_SECRET, registry: env.DOCKER_REGISTRY)
-              script {
-                getVaultSecret.readSecretWrapper {
-                  sh(label: 'release-manager.sh', script: '.ci/scripts/release-manager.sh')
-                }
-              }
+            releaseManager(type: 'snapshot')
+            whenFalse(env.BRANCH_NAME.equals('main')) {
+              releaseManager(type: 'staging')
             }
```
Comment (v1v) [inline .ci/packaging.groovy:115]: Run always `snapshot` for `main`/`8.1`/`8.2` branches and also `staging` for `8.1`/`8.2` branches

File: .ci/packaging.groovy [hunk 2]
```diff
@@ -154,6 +128,61 @@ pipeline {
...
+      credentialsId: "${JOB_GCS_CREDENTIALS}",
+      pathPrefix: "${BASE_DIR}/build/",
+      pattern: "${BASE_DIR}/build/dependencies.csv",
+      sharedPublicly: true,
+      showInline: true)
+  }
+}
+
+def getBucketLocation(type) {
+  return "gs://${JOB_GCS_BUCKET}/${URI_SUFFIX}/${type}"
```
Comment (v1v) [inline .ci/packaging.groovy:183]: Artifacts are not stored in either the `staging` or `snapshot` subfolders, this is quite handy in order to run the same workflow for different type of releases

File: .ci/packaging.groovy [hunk 3]
```diff
@@ -79,6 +78,10 @@ pipeline {
                 name 'PLATFORM'
                 values 'linux && immutable', 'arm'
               }
+              axis {
+                name 'TYPE'
+                values 'snapshot', 'staging'
```
Comment (axw) [inline .ci/packaging.groovy:83]: is it possible to also exclude staging here when on main?

**Review by axw — APPROVED** [review]
Looks great! Just one comment - I think it would be best to avoid running `make release-manager-release` on the main branch, as it's quite an expensive target.

File: .ci/packaging.groovy (hunk 3 above)
Comment (v1v) [reply .ci/packaging.groovy:83]: I tried to use https://github.com/elastic/apm-server/pull/7584/commits/eaa81b1b0ddaeb10462e7832c5f186bd78f11ac4 (the declarative form) but somehow the UI representation skipped one of the stages.

image

So I finally decided to use the scripted approach with https://github.com/elastic/apm-server/pull/7584/commits/e46eb9bac19a6baffdb2feeff223be59062a1b83

image

**Review by cachedout — APPROVED** [review]

**Pull request merged**